File: game_plugins/piranhas_2019/shared/sc/plugin2019/Board.java
package sc.plugin2019;

import com.thoughtworks.xstream.annotations.XStreamAlias;
import com.thoughtworks.xstream.annotations.XStreamImplicit;
import sc.plugin2019.util.Constants;
import sc.shared.PlayerColor;

import static sc.plugin2019.FieldState.EMPTY;
import static sc.plugin2019.FieldState.OBSTRUCTED;

/**
 * Ein Spielbrett bestehend aus 8x8 Feldern
 */
@XStreamAlias(value = "board")
public class Board {

  @XStreamImplicit(itemFieldName = "fields")
  private Field[][] fields;

  public Board() {
    this.fields = new Field[Constants.BOARD_SIZE][Constants.BOARD_SIZE];
    initialize();
  }

  /**
   * Erstellt eine zufällige Spielbrett. Dazu werden
   */
  private void initialize() {
    for (int x = 0; x < Constants.BOARD_SIZE; x++) {
      for (int y = 0; y < Constants.BOARD_SIZE; y++) {
        fields[x][y] = new Field(x,y);
      }
    }
    // place piranhas
    for (int index = 1; index < Constants.BOARD_SIZE - 1; index++) {
      fields[0][index].setPiranha(PlayerColor.RED);
      fields[Constants.BOARD_SIZE -1][index].setPiranha(PlayerColor.RED);
      fields[index][0].setPiranha(PlayerColor.BLUE);
      fields[index][Constants.BOARD_SIZE -1].setPiranha(PlayerColor.BLUE);
    }
    int firstX = 0, firstY = 0;
    for(int i = 0; i < Constants.NUM_OBSTICLES; i++){
      int x,y;

      // Generate x y coordinate on empty field
      // obstructed fields are in the inner 6x6 field and are not allowed to be in same vertical, diagonal or horizontal line
      do{
        x = (int) (Math.random()*6+2);
        y = (int) (Math.random()*6+2);
      } while(fields[x][y].getState() != EMPTY &&
              (i == 0 || // if first field was generated
                      (firstX == x || firstY == y || // check the generation conditions for second field
                        firstX - firstY == x - y ||
                        firstX + firstY == x + y)));

      fields[x][y].setState(OBSTRUCTED);
      firstX = x;
      firstY = y;
    }

  }

  /**
   * erzeugt eine Deepcopy dieses Objekts
   *
   * @return ein neues Objekt mit gleichen Eigenschaften
   */
  @Override
  public Board clone() {
    Board clone = new Board();
    for (int x = 0; x < Constants.BOARD_SIZE; x++) {
      for (int y = 0; y < Constants.BOARD_SIZE; y++) {
        clone.fields[x][y] = fields[x][y].clone();
      }
    }
    return clone;
  }

  @Override
  public String toString() {
    String toString = "Board:\n";
    StringBuilder b = new StringBuilder(toString);
    for (int x = 0; x < Constants.BOARD_SIZE; x++) {
      for (int y = 0; y < Constants.BOARD_SIZE; y++) {
        b.append(fields[x][y].getPiranha());
      }
    }
    return b.toString();
  }

  public Field getField(int x, int y) {
    return this.fields[x][y];
  }
}

File: plugin2099/src/main/kotlin/sc/plugin2099/Board.kt
package sc.plugin2099

import com.thoughtworks.xstream.annotations.XStreamAlias
import com.thoughtworks.xstream.annotations.XStreamImplicit
import sc.api.plugins.*
import sc.framework.deepCopy
import sc.plugin2099.util.TicTacToeConstants

val line = "-".repeat(TicTacToeConstants.BOARD_LENGTH * 2 + 2)

/** Spielbrett für Piranhas mit [TicTacToeConstants.BOARD_LENGTH]² Feldern.  */
@XStreamAlias(value = "board")
class Board(
    @XStreamImplicit(itemFieldName = "row")
    override val gameField: MutableTwoDBoard<FieldState> = emptyFields()
): RectangularBoard<FieldState>(), IBoard {
    
    override fun toString() =
        "Board " + gameField.withIndex().joinToString(" ", "[", "]") { row ->
            row.value.withIndex().joinToString(", ", prefix = "[", postfix = "]") {
                "(${it.index}, ${row.index}) " + it.value.toString()
            }
        }
    
    fun prettyString(): String {
        val map = StringBuilder(line)
        gameField.forEach { row ->
            map.append("\n|")
            row.forEach { field ->
                map.append(field.asLetters())
            }
        }
        map.append("\n").append(line)
        return map.toString()
    }
    
    override fun clone(): Board {
        //println("Cloning with ${gameField::class.java}: $this")
        return Board(gameField.deepCopy())
    }
    
    fun getTeam(pos: Coordinates): Team? =
        this[pos].team
    
    
    companion object {
        /** Erstellt ein leeres Spielbrett.  */
        fun emptyFields(): MutableTwoDBoard<FieldState> {
            return Array(TicTacToeConstants.BOARD_LENGTH) {
                Array(TicTacToeConstants.BOARD_LENGTH) { FieldState.EMPTY }
            }
        }
    }
}
